pkg/lib/cockpit/channel.ts

```diff
@@ -0,0 +1,240 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { JsonObject } from './_internal/common';
+import { Transport, ensure_transport, transport_globals } from './_internal/transport';
+import { EventEmitter } from './event';
+
+type ChannelPayload = string | Uint8Array;
+
+type ChannelBinaryOption<P extends ChannelPayload> =
+    P extends Uint8Array ?
+        { binary: true }
+    : P extends string ?
+        { binary?: false }
+    :
+        { binary: boolean }
+    ;
+
+type ChannelOptions<P extends ChannelPayload> = JsonObject & ChannelBinaryOption<P> & {
+    command?: never;
+    channel?: never;
+    host?: string;
+    payload: string;
+    superuser?: "try" | "require";
+};
+
+interface ChannelControlMessage extends JsonObject {
+    command: string;
+}
+
+interface ChannelEvents<P extends ChannelPayload> {
+    control(options: ChannelControlMessage): void;
+    done(options: ChannelControlMessage): void;
+    ready(options: ChannelControlMessage): void;
+    close(options: ChannelControlMessage): void;
+    data(data: P): void;
+}
```

Comment:
I'm a bit surprised that there's no `ChannelError` error type anywhere. Indeed this abstract Channel gets away with not having it, as it just passes through `problem` and `options` in `close`, and treats everything else as a generic `JsonObject`. Would this be a good place to define an error type and use that instead of `JsonObject` here? For typing `.problem`, `.message`, `.host-key` etc? (Not needed in this PR, I'm just wondering).

Reply:
`close` can also contain non-error type of things like `tag`, and in that case `problem` wouldn't be defined.

I suppose we could have something like:

```
interface CloseControlMessage {
  problem?: string;
  ....etc
}
```

where everything is optional.  I wouldn't mind to include the attributes attached to errors thrown by the bridge in that case (hostkey, etc), but again, all optional.  Would that help you?

I feel like if we did have a `CockpitProblem` type of interface, it could then be a subclass of this interface which has `problem` strictly defined as string.  After all, we only throw if `problem` is set.

Reply:
I don't yet know if it would help me -- I suppose that depends on your reaction to #20726 :grin: Given that caught exceptions are always `unknown`, I'm happy to live with run-time property tests and duck typing -- we know how to do that, and I don't think we can do statically typed errors at this point.

I.e. this was less of a "change this" and more of poking your brain about where you want this to go.

Reply:
Let's leave this unaddressed for now, then.